### PR TITLE
ENH: no_degenerates operates on array data

### PR DESCRIPTION
**Pull request by KatherineCaley**
## Summary by Sourcery

Add a new method 'from_names_and_array' to the AlignedSeqsData class for constructing objects from names and array data, and refactor the 'no_degenerates' method to enhance performance by operating on array data. Update tests to cover new functionality and edge cases.

New Features:
- Introduce a new class method 'from_names_and_array' in the AlignedSeqsData class to construct an object from a list of names and a numpy array of aligned sequence data.

Enhancements:
- Refactor the 'no_degenerates' method to operate on array data, improving performance and clarity.

Tests:
- Add new tests for the 'from_names_and_array' method to ensure correct functionality, including cases for empty data and mismatched lengths.

### CODE REVIEW

File: tests/test_core/test_new_alignment.py [hunk 1]
```diff
@@ -2804,6 +2804,38 @@ def test_from_seqs_and_maps_diff_keys_raises(dna_alphabet):
...
+def test_from_names_and_array(dna_alphabet):
+    names = ["seq1", "seq2", "seq3"]
+    data = numpy.array([[0, 1, 2, 3], [3, 2, 1, 0], [4, 4, 4, 4]])
+    asd = new_alignment.AlignedSeqsData.from_names_and_array(
+        names=names, data=data, alphabet=dna_alphabet
+    )
+    assert asd.names == ["seq1", "seq2", "seq3"]
+    got = asd.get_seq_array(seqid="seq1")
+    expect = numpy.array([0, 1, 2, 3])
+    assert numpy.array_equal(got, expect)
```
[the remark this answers is not in the document]
Comment (KatherineCaley) [reply tests/test_core/test_new_alignment.py:2816]: very ironic considering your advice to avoid loops in tests!!

File: src/cogent3/core/new_alignment.py [hunk 2]
```diff
@@ -2955,6 +2965,43 @@ def from_seqs_and_gaps(
...
+            alphabet object for the sequences
+        """
+        if len(names) != data.shape[0]:
+            raise ValueError("Number of names must match number of rows in data.")
+
+        seqs = {}
+        gaps = {}
+        for name, seq in zip(names, data):
+            seq, gap_map = seq_to_gap_coords(seq, alphabet=alphabet)
+            seq.flags.writeable = False
```
[the remark this answers is not in the document]
Comment (KatherineCaley) [reply src/cogent3/core/new_alignment.py:2995]: array is constructed in seq_to_gap_coords

File: src/cogent3/core/new_alignment.py [hunk 3]
```diff
@@ -4333,7 +4380,6 @@ def no_degenerates(self, motif_length: int = 1, allow_gap: bool = False):
             character (latter is default, as most evolutionary modelling treats
             gaps as N).
         """
-        # refactor: array
         try:
             chars = list(self.moltype.alphabet)
         except AttributeError as e:
```
Comment (GavinHuttley) [inline src/cogent3/core/new_alignment.py:4385]: I think every new moltype has an `.alphabet` attribute. Change this check to
```python
if self.moltype.degen_alphabet is None:
    raise MolTypeError(...)

chars = len(self.moltype)
```
Note I've also changed the exception type. We need a custom `MolTypeError` (which inherits from TypeError). That's what should be raised when, for example, someone tries to reverse complement a sequence that is not DNA or RNA.

Also, we need a test that triggers this condition.

File: src/cogent3/core/new_alignment.py [hunk 4]
```diff
@@ -4343,14 +4389,27 @@ def no_degenerates(self, motif_length: int = 1, allow_gap: bool = False):
             )
             raise ValueError(msg) from e
 
-        if allow_gap:
-            chars.extend(self.moltype.gap)
+        array_pos = self.array_positions
+        cutoff = len(chars) + 1 if allow_gap else len(chars)
```
Comment (GavinHuttley) [inline src/cogent3/core/new_alignment.py:4393]: add a comment before this definition of cutoff stating that, by design, char alphabets are organised such that the canonical characters always occur first, followed by gap, then ambiguity characters.

File: src/cogent3/core/new_alignment.py [hunk 5]
```diff
@@ -4343,14 +4389,27 @@ def no_degenerates(self, motif_length: int = 1, allow_gap: bool = False):
...
+        indices = (array_pos < cutoff).all(axis=1)
 
-        def predicate(column):
-            data = set("".join([str(d) for d in column]))
-            return data <= set(chars)
+        if motif_length > 1:
+            num_motif = len(self) // motif_length
 
-        return self.filtered(predicate, motif_length=motif_length)
+            if len(self) % motif_length != 0:
```
Comment (GavinHuttley) [inline src/cogent3/core/new_alignment.py:4399]: simplify this
```python
if remainder := len(self) % motif_length:
    indices = indices[:-remainder]
    ...
```

**Review by GavinHuttley — CHANGES_REQUESTED** [review]
some suggestions for improved errors plus checking moltypes etc..

File: src/cogent3/core/new_alignment.py [hunk 6]
```diff
@@ -4380,24 +4380,26 @@ def no_degenerates(self, motif_length: int = 1, allow_gap: bool = False):
...
+        cutoff = chars + 1 if allow_gap else chars
         indices = (array_pos < cutoff).all(axis=1)
 
         if motif_length > 1:
             num_motif = len(self) // motif_length
 
-            if len(self) % motif_length != 0:
-                indices = indices[: num_motif * motif_length]
+            if remainder := len(self) % motif_length:
+                indices = indices[:-remainder]
```
Comment (GavinHuttley) [inline src/cogent3/core/new_alignment.py:4402]: you can probably delete `num_pos` and just slice array_pos the same as indices, i.e. `array_pos = array_pos[:-remainder]`

**Review by GavinHuttley — APPROVED** [review]
But I do suggest a small simplification...

File: src/cogent3/core/new_alignment.py (hunk 6 above)
Comment (KatherineCaley) [reply src/cogent3/core/new_alignment.py:4402]: I'll make the change and then merge
